### DISCUSSION TITLE
Constraint on time derivatives

### DISCUSSION THREAD

**Discussion by mrrezaie** [casadi/casadi#3562] [Q&A] [answered]
Hi, I have two matrix variables and want to define a constraint that the first time derivative of A equals B. The very first way is to use dAdt==B, but I will miss the last row:
```python
opti.subject_to( (A[1:,:]-A[:-1,:])/dt == B[:-1,:] )
```
Is there any better way of doing this?

Outside casadi, I could use `np.gradient(A, dt, edge_order=2)` which gave me same shape. Although I could replicate this, but wanted to know if there are some built-in functions in this regard.

Thanks in advance.

**Comment by sandeep026** [comment, marked as answer]
Numpy's gradient uses 2nd order approximation at interior point and first order approximation at boundaries. It can be implemented easily as follows.
Let c and t be two vectors of length n+1.

```
cdot=
cs.horcat((c[:,1]-c[:,0])/(t[:,1]-t[:,0]),(c[:,0:n-1]-c[:,2:n+1])/(2*(t[:,1:n+1] - t[:,0:n])),(c[:,n]-c[:,n-1])/(t[:,n]-t[:,n-1]))

```

**Reply by sandeep026** [reply]
Note:  ```opti.subject_to( (A[1:,:]-A[:-1,:]/dt) == B[:-1,:] )``` should be converted to ```opti.subject_to( (A[1:,:]-A[:-1,:]) == B[:-1,:]dt )``` to avoid Nan detected while solving the optimal control problems.

**Reply by mrrezaie** [reply]
Thanks @sandeep026 for your response. You are correct, I made a mistake in syntaxing. Could you please elaborate a little bit why Nans may be detected if I set `dt` as denominator? 
Indeed, I was looking for something without indexing. Do know if any is available? I have more constraints with both variables and want them to be with same shape. Thanks for your help.

**Reply by sandeep026** [reply]
If dt is also part of your optimization problem, being in the denominator, it can become zero and thereby the constraint blows up and reported as nan by solver.

**Reply by mrrezaie** [reply]
Oh yes, the time interval is a constant, positive, and non-zero value (0.01 s in my case). Thank you.
In Pyomo, I could define a `Var` and its derivative, `DerivativeVar`, and define a constraint without explicit formulation. Does CasADi have equivalent feature?

**Reply by sandeep026** [reply]
Nope. Casadi is meant to provide building blocks for your optimal control application.

**Reply by jgillis** [reply]
@sandeep026 is right. Perhaps other tools built on top of CasADi suit your needs better: see bottom page of casadi.org - in particular have a look at openocl, rockit, yop and nmpyc

**Reply by mrrezaie** [reply]
Thanks for your help.
